tokio-util/src/codec/framed_impl.rs

```diff
@@ -122,13 +122,20 @@ where
         let state: &mut ReadFrame = pinned.state.borrow_mut();
         loop {
             // Repeatedly call `decode` or `decode_eof` as long as it is
-            // "readable". Readable is defined as not having returned `None`. If
-            // the upstream has returned EOF, and the decoder is no longer
-            // readable, it can be assumed that the decoder will never become
-            // readable again, at which point the stream is terminated.
+            // "readable". Readable is defined as not having returned `None`.
+            // If the upstream has returned EOF, and the decoder is no longer
+            // readable, it might become readable again if the underlying
+            // AsyncRead is a FIFO or file.
+            // However, we still want to make sure that upon encountering an EOF
+            // we actually finish emmiting all it's associated decode_eof frames.
+            // Furthermore, we don't want to emit any decode_eof frames on retried
+            // reads after an EOF unless we've actually read more data.
```

Comment:
Can we include the ascii drawing somewhere in this file for future readers?

Reply:
I was already on that 😄!